java/org/contikios/cooja/Simulation.java

```diff
@@ -150,68 +155,133 @@ public Simulation(Cooja cooja, long seed) {
     this.cooja = cooja;
     randomGenerator = new SafeRandom(this);
     randomSeed = seed;
-  }
-
-  @Override
-  public void run() {
-    assert isRunning : "Did not set isRunning before starting";
-    lastStartRealTime = System.currentTimeMillis();
-    lastStartSimulationTime = getSimulationTimeMillis();
-    speedLimitLastRealtime = lastStartRealTime;
-    speedLimitLastSimtime = lastStartSimulationTime;
+    this.simulationThread = new Thread(() -> {
+      boolean isSimulationRunning = false;
+      boolean isAlive = true;
+      do {
+        EventQueue.Pair nextEvent = null;
+        try {
+          while (isAlive) {
+            Runnable cmd;
+            do {
+              cmd = isSimulationRunning ? commandQueue.poll() : commandQueue.take();
+              if (cmd == null) {
+                // No command to execute
+              } else if (cmd == START) {
+                if (!isSimulationRunning) {
+                  isSimulationRunning = true;
+                  setRunning(true);
+                }
+              } else if (cmd == STOP) {
+                if (isSimulationRunning) {
+                  isSimulationRunning = false;
+                  setRunning(false);
+                }
+              } else if (cmd == QUIT) {
+                isAlive = false;
+                isShutdown = true;
+                if (isSimulationRunning) {
+                  isSimulationRunning = false;
+                  setRunning(false);
+                }
+                // Process no more commands
+                break;
+              } else {
+                cmd.run();
+              }
+            } while (cmd != null);
+
+            if (isSimulationRunning) {
```

Comment:
Can this be reached from anything other than `cmd == START` and `cmd == QUIT`? Could the condition be `if (isAlive) {` instead?

Reply:
Yes, it could be `isAlive` instead of `isSimulationRunning` but is that any advantage in changing?

Reply:
Wouldn't that enable the removal of `isSimulationRunning`?

Reply:
Well, `isSimulationRunning` is used to decide if the simulation should `poll` or `take` from the command queue. We could use the volatile global variable `isRunning` for this but I wanted to avoid accessing it from the inner event loop if possible.

Reply:
Using `isAlive` here allows the scope to be reduced for `isSimulationRunnning` (to same scope as `Object cmd`).